// react-i18nify/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "I18n", {
  enumerable: true,
  get: function () {
    return _I18n.default;
  }
});
Object.defineProperty(exports, "Localize", {
  enumerable: true,
  get: function () {
    return _Localize.default;
  }
});
Object.defineProperty(exports, "Translate", {
  enumerable: true,
  get: function () {
    return _Translate.default;
  }
});
Object.defineProperty(exports, "forceComponentsUpdate", {
  enumerable: true,
  get: function () {
    return _Base.forceComponentsUpdate;
  }
});
Object.defineProperty(exports, "getLocale", {
  enumerable: true,
  get: function () {
    return _settings.getLocale;
  }
});
Object.defineProperty(exports, "getTranslations", {
  enumerable: true,
  get: function () {
    return _settings.getTranslations;
  }
});
Object.defineProperty(exports, "l", {
  enumerable: true,
  get: function () {
    return _localize.default;
  }
});
Object.defineProperty(exports, "localize", {
  enumerable: true,
  get: function () {
    return _localize.default;
  }
});
Object.defineProperty(exports, "setHandleFailedLocalization", {
  enumerable: true,
  get: function () {
    return _settings.setHandleFailedLocalization;
  }
});
Object.defineProperty(exports, "setHandleMissingTranslation", {
  enumerable: true,
  get: function () {
    return _settings.setHandleMissingTranslation;
  }
});
Object.defineProperty(exports, "setLocale", {
  enumerable: true,
  get: function () {
    return _settings.setLocale;
  }
});
Object.defineProperty(exports, "setLocaleGetter", {
  enumerable: true,
  get: function () {
    return _settings.setLocaleGetter;
  }
});
Object.defineProperty(exports, "setTranslations", {
  enumerable: true,
  get: function () {
    return _settings.setTranslations;
  }
});
Object.defineProperty(exports, "setTranslationsGetter", {
  enumerable: true,
  get: function () {
    return _settings.setTranslationsGetter;
  }
});
Object.defineProperty(exports, "t", {
  enumerable: true,
  get: function () {
    return _translate.default;
  }
});
Object.defineProperty(exports, "translate", {
  enumerable: true,
  get: function () {
    return _translate.default;
  }
});
Object.defineProperty(exports, "translateReplace", {
  enumerable: true,
  get: function () {
    return _utils.replace;
  }
});
var _Base = require("./components/Base.js");
var _Translate = _interopRequireDefault(require("./components/Translate.js"));
var _Localize = _interopRequireDefault(require("./components/Localize.js"));
var _I18n = _interopRequireDefault(require("./components/I18n.js"));
var _settings = require("./lib/settings");
var _translate = _interopRequireDefault(require("./lib/translate"));
var _localize = _interopRequireDefault(require("./lib/localize"));
var _utils = require("./lib/utils");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
